# EDA

«Предсказать поведение, чтобы удержать клиентов. Вы можете проанализировать все соответствующие данные о клиентах и разработать целенаправленные программы удержания клиентов». [Наборы Данных Образца IBM]

**Объекты:**

Каждая строка представляет клиента, каждый столб содержит атрибуты клиента, описанные в столбце Metadata.

**Набор данных включает информацию о:**

- Клиенты, которые ушли в течение последнего месяца – колонка называется Churn
- Услуги, на которые подписался каждый клиент - телефон, несколько линий, интернет, онлайн-безопасность, онлайн-резервное копирование, защита устройств, техническая поддержка и потоковое телевидение и фильмы
- Информация об учетной записи клиента - как долго он был клиентом, контракт, способ оплаты, безбумажный выставление счетов, ежемесячные платежи и общие расходы
- Демографическая информация о клиентах – пол, возрастной диапазон, а также о том, есть ли у них партнеры и иждивенцы

Датасет взят из [Telco Customer Churn](https://www.kaggle.com/datasets/blastchar/telco-customer-churn?select=WA_Fn-UseC_-Telco-Customer-Churn.csv)

**Цель ноутбука:** первичное исследование данных перед feature engineering.


## Импорт библиотек

In [1]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt

## Загрузка данных

In [28]:
df = pd.read_csv('../data/raw/churn.csv')
print(df.shape)
print(df.dtypes)
df.head()

(7043, 21)
customerID              str
gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64
TotalCharges            str
Churn                   str
dtype: object


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


Исследуемый датасет имеет 7043 строки (объекта) с 21 признаком.  
Каждый признак имеет тип str, int64 или float64, следовательно в каждом столбце данные одного типа.  
Среди признаков найден TotalCharges имееющий тип str, хотя фактически в нем записаны числа.  

## Проверка целостности данных

In [38]:
print(f'Наличие дубликатов: {df.duplicated().any()}')
print(f'Дубликаты среди customerId: {df['customerID'].duplicated().any()}')
print(f'Наличие пропусков (NaN):\n{df.isna().sum()}')
df['TotalCharges_num'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
df.loc[df['TotalCharges_num'].isna()]

Наличие дубликатов: False
Дубликаты среди customerId: False
Наличие пропусков (NaN):
customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,TotalCharges_num
488,4472-LVYGI,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,No,Two year,Yes,Bank transfer (automatic),52.55,,No,NaN
753,3115-CZMZD,Male,0,No,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.25,,No,NaN
936,5709-LVOEQ,Female,0,Yes,Yes,0,Yes,No,DSL,Yes,...,No,Yes,Yes,Two year,No,Mailed check,80.85,,No,NaN
1082,4367-NUYAO,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.75,,No,NaN
1340,1371-DWPAZ,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,No,Two year,No,Credit card (automatic),56.05,,No,NaN
3331,7644-OMVMY,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.85,,No,NaN
3826,3213-VVOLG,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.35,,No,NaN
4380,2520-SGTTA,Female,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.00,,No,NaN
5218,2923-ARZLG,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,One year,Yes,Mailed check,19.70,,No,NaN
6670,4075-WKNIU,Female,0,Yes,Yes,0,Yes,Yes,DSL,No,...,Yes,Yes,No,Two year,No,Mailed check,73.35,,No,NaN


При переводе в числовой тип признак TotalCharges появились NaN у 11 объектов, у всех 11 объектов tenure (срок прибывания) равен 0, следовательно данные клиенты еще не успели совершить транзакций, поэтому их общие траты будет разумно прировнять к 0. Помимо этого у всех стоит churn=0, что означает, что ни один из них пока не ушел, что логично, ведь данные клиенты только подключились